### PR TITLE
Simplify `min_match_length`

### PR DESCRIPTION
**Pull request by calleerlandsson**
Simplify `min_match_length` by replacing the existing iterative
algorithm with a recursive one.

### CODE REVIEW

**Comment by calleerlandsson** [comment]
@mike-burns, @mptre I would love your feedback on this.

**Comment by mptre** [comment]
I like the idea, however the proposed implementation does not fully utilize the recursive aspect since the computation of the `match_length` function is better off solved using iteration. Instead I would remove the `match_length` auxiliary function and let `min_match_length` do the minimization. Here's a rough outline of my proposal:

```c
static size_t
min_match_length(char *string)
{
	char *start, *end;
	size_t n, m;
	int i;

	start = strchr(string, query[0]);
	if (start == NULL)
		return INT_MAX;

	for (i = 1, end = start + 1; *end && query[i]; end++)
		if (*end == query[i])
			i++;
	n = !query[i] ? end - start : INT_MAX;
	m = min_match_length(string + 1);

	return n < m ? n : m;
}
```

`0` can't be used as a indicator of a missing match since we want to minimize the length therefore `INT_MAX` is used instead. The condition in `score` needs to be adjusted as well:

```c
if ((match_length = min_match_length(string)) == INT_MAX)
```

**Comment by calleerlandsson** [comment]
@mptre That's a great proposal: Makes it even simpler! I have pushed a commit with your proposed algorithm.

> the proposed implementation does not fully utilize the recursive aspect since the computation of the match_length function is better off solved using iteration

Because I'm curious and would love to learn, would you mind elaborating on this? Why is this better off solved using iteration?

**Comment by mptre** [comment]
> Because I'm curious and would love to learn, would you mind elaborating on
> this? Why is this better off solved using iteration?

I can at least try: the prior implementation could easily be solved with a
for-loop in conjunction with a nested for-loop. The overhead of the recursive
call is therefore not justified in my opinion. My proposal could also be solved
with a for-loop but I find using recursion for (min|max)imization problems very
elegant and intuitive since the right answer will "bubble" up in the call stack.

On Tue, Jan 05, 2016 at 01:55:31AM -0800, Calle Erlandsson wrote:
> @mptre That's a great proposal: Makes it even simpler! I have pushed a commit with your proposed algorithm.
> 
> > the proposed implementation does not fully utilize the recursive aspect since the computation of the match_length function is better off solved using iteration
> 
> Because I'm curious and would love to learn, would you mind elaborating on this? Why is this better off solved using iteration?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/thoughtbot/pick/pull/83#issuecomment-168955132

-- 
:wq

File: src/main.c [hunk 1]
```diff
@@ -479,38 +481,33 @@ score(char *string)
...
-			if (match_length == 0 || match_length > match_position -
-			    match_start + 1)
-				match_length = match_position - match_start + 1;
-		}
-	}
+	char	*start, *end;
+	int	 i;
+	size_t	 length, min;
+
+	if (!(start = strcasechr(string, query[0])))
```
Comment (mptre) [inline src/main.c]: Why not use `strncasecmp`?

```c
if (!(start = strncasecmp(string, query, 1)))
```

File: src/main.c [hunk 2]
```diff
@@ -479,38 +481,33 @@ score(char *string)
...
+		return INT_MAX;
+	for (i = 1, end = start + 1; *end && query[i]; end++)
+		if (tolower((unsigned char)*end) == tolower((unsigned
```
Comment (mptre) [inline src/main.c]:
```c
if (strncasecmp(end, &query[i], 1) == 0)
```

File: src/main.c (hunk 1 above)
Comment (calleerlandsson) [inline src/main.c]: In this case, `start` has to be a `char *` to the start of the match but `strncasecmp` returns an `int`. I have updated `strcasechr` to use `strncasecmp`, though in af25c5f.

File: src/main.c (hunk 2 above)
Comment (calleerlandsson) [inline src/main.c]: Good catch! I didn't think about that. Fixed in af25c5f.

File: src/main.c (hunk 1 above)
Comment (mptre) [inline src/main.c]: Ah, my bad.

On Thu, Jan 07, 2016 at 02:14:32AM -0800, Calle Erlandsson wrote:
> > -					if (tolower((unsigned char)string[match_position]) == query_char) {
> > -						++match_position;
> > -						break;
> > -					}
> > -				}
> > -			}
> > -			if (match_length == 0 || match_length > match_position -
> > -			    match_start + 1)
> > -				match_length = match_position - match_start + 1;
> > -		}
> > -	}
> > +	char	*start, *end;
> > +	int	 i;
> > +	size_t	 length, min;
> > +
> > +	if (!(start = strcasechr(string, query[0])))
> 
> In this case, `start` has to be a `char *` to the start of the match but `strncasecmp` returns an `int`. I have updated `strcasechr` to use `strncasecmp`, though in af25c5f.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/thoughtbot/pick/pull/83/files#r49058819

-- 
:wq

**Comment by calleerlandsson** [comment]
Thanks for the review, @mptre. I'll merge squash and merge this.

**Pull request merged**